parse-zoneinfo/src/line.rs

```diff
@@ -628,12 +623,64 @@ pub enum TimeType {
     UTC,
 }
 
+impl TimeType {
+    fn from_char(c: char) -> Option<Self> {
+        Some(match c {
+            'w' => Self::Wall,
+            's' => Self::Standard,
+            'u' | 'g' | 'z' => Self::UTC,
+            _ => return None,
+        })
+    }
+}
+
 #[derive(PartialEq, Debug, Copy, Clone)]
 pub struct TimeSpecAndType(pub TimeSpec, pub TimeType);
 
-impl TimeSpec {
-    pub fn with_type(self, timetype: TimeType) -> TimeSpecAndType {
-        TimeSpecAndType(self, timetype)
+impl FromStr for TimeSpecAndType {
+    type Err = Error;
+
+    fn from_str(input: &str) -> Result<Self, Error> {
+        if input == "-" {
+            return Ok(TimeSpecAndType(TimeSpec::Zero, TimeType::Wall));
+        } else if input.chars().all(|c| c == '-' || c.is_ascii_digit()) {
+            return Ok(TimeSpecAndType(
+                TimeSpec::Hours(input.parse().unwrap()),
+                TimeType::Wall,
+            ));
+        }
+
+        let (input, ty) = match input.chars().last().and_then(TimeType::from_char) {
+            Some(ty) => (&input[..input.len() - 1], Some(ty)),
+            None => (input, None),
+        };
+
+        let neg = if input.starts_with('-') { -1 } else { 1 };
+        let mut state = TimeSpec::Zero;
+        for part in input.split(':') {
+            state = match (state, part) {
+                (TimeSpec::Zero, hour) => TimeSpec::Hours(
+                    i8::from_str(hour)
+                        .map_err(|_| Error::InvalidTimeSpecAndType(input.to_string()))?,
+                ),
+                (TimeSpec::Hours(hours), minutes) => TimeSpec::HoursMinutes(
+                    hours,
+                    i8::from_str(minutes)
```

Comment:
The regex checked minutes and seconds had two digits to prevent parsing `1:1`, `1:1:1` or `1:00000` as a valid time. Can you add a length check? (Maybe also worth a test)